### PR TITLE
Uart/Console driver for hotel

### PR DESCRIPTION
**Pull request by alevy**
Adds an implementation of the `UART` trait for the hotel chip's UART, and uses that to plug it into the console driver for Tango and Golf platforms.

The UART trait implementation is technically incomplete (there are several methods implemented with `unimplemented`), _but_ those methods should be removed from the trait (they allow synchronous reading/writing) and aren't used anywhere in any port of the OS.

To do this, we also needed to add the input direction for the Pinmux controller.

### CODE REVIEW

File: src/chips/hotel/src/interrupts.rs [hunk 1]
```diff
@@ -175,10 +175,10 @@ pub static INTERRUPT_TABLE: [Option<unsafe extern fn()>; 203] =
       None, // TRNG0_INTR_READ_EMPTY_INT
       None, // UART0_RXBINT
       None, // UART0_RXFINT
-      None, // UART0_RXINT
+      Some(::uart::uart0_rx_handler), // UART0_RXINT
```
Comment (yuriks) [inline src/chips/hotel/src/interrupts.rs]: Shouldn't these be handler for the other UARTs too?

File: src/chips/hotel/src/interrupts.rs [hunk 2]
```diff
@@ -175,10 +175,10 @@ pub static INTERRUPT_TABLE: [Option<unsafe extern fn()>; 203] =
...
       None, // UART0_RXOVINT
       None, // UART0_RXTOINT
-      Some(::uart::uart0_handler), // UART0_TXINT
+      Some(::uart::uart0_tx_handler), // UART0_TXINT
```
Comment (yuriks) [inline src/chips/hotel/src/interrupts.rs]: Shouldn't these be handler for the other UARTs too?

File: src/chips/hotel/src/chip.rs [hunk 3]
```diff
@@ -27,8 +28,14 @@ impl Chip for Hotel {
         unsafe {
             cortexm3::nvic::next_pending().map(|nvic_num| {
                 match nvic_num {
-                    _ => {}
+                    174 => uart::UART0.handle_rx_interrupt(),
+                    177 => uart::UART0.handle_tx_interrupt(),
+                    184 => uart::UART1.handle_tx_interrupt(),
+                    191 => uart::UART2.handle_tx_interrupt(),
```
Comment (yuriks) [inline src/chips/hotel/src/chip.rs]: What about other rx interrupts?

File: src/chips/hotel/src/uart.rs [hunk 4]
```diff
@@ -127,6 +133,36 @@ impl UART {
...
+        self.clock.enable();
+
+        let ctrl = regs.control.get() | 0b10;
+        regs.control.set(ctrl);
+        regs.interrupt_control.set(regs.interrupt_control.get() | 2);
+    }
+
+    /// Disable reception on the UART
+    ///
+    /// Side-effect: turns the clock off if RX is also disabled.
```
Comment (yuriks) [inline src/chips/hotel/src/uart.rs]: "if TX* is also disabled"

File: src/chips/hotel/src/uart.rs [hunk 5]
```diff
@@ -127,6 +133,36 @@ impl UART {
...
+    ///
+    /// Side-effect: ensures the clock is on.
+    pub fn enable_rx(&self) {
+        let regs = unsafe { &*self.regs };
+
+        self.clock.enable();
+
+        let ctrl = regs.control.get() | 0b10;
+        regs.control.set(ctrl);
+        regs.interrupt_control.set(regs.interrupt_control.get() | 2);
```
Comment (yuriks) [inline src/chips/hotel/src/uart.rs]: Why don't you need to enable/disabled the tx interrupt?

File: src/chips/hotel/src/uart.rs [hunk 6]
```diff
@@ -234,11 +266,42 @@ impl UART {
...
         if self.send_remaining_bytes() == 0 {
-            let _userbuf = self.buffer.take();
-            // Done sending, return buffer to client
+            self.client.map(|client| {
+                self.buffer.take().map(move |buffer| {
+                    match buffer.bytes {
+                        EitherBytes::Mutable(bytes) => {
+                            client.write_done(bytes)
+                        },
+                        _ => {}
```
Comment (yuriks) [inline src/chips/hotel/src/uart.rs]: This should do something/panic, I think

**Comment by yuriks** [comment]
Overall LGTM. One thing I've been struggling to understand is the point of `TakeCell`. Most uses of it I've seen so far it's been acting as essentially a less-ergonomic `RefCell` that instead of complaining loudly if you have a bug will just silently do nothing and leave you with incorrect/partial results. This isn't something specific to the PR, though you are using TakeCell, which is why I'm bringing it up.

**Comment by alevy** [comment]
@yuriks Address you're comments.

Particularly of note, I just removed the `EitherBytes` type since it's not actually useful for existing clients. The thought is to balance allowing clients to write immutable buffers like `&str`s while being able to return a mutable version when that's what was passed in. But for that to be at all useful, it has to bubble up to the client.

Regarding TakeCell, the point is exactly to avoid panics at the expense of functionality bugs (that are isolated to individual components). We want to think of each component as being potentially buggy. If the UART, say, is buggy and but doesn't panic, maybe bytes don't go over the wire for a bit but everything else continues on. If the UART is buggy and therefore panics, it means the whole system goes down or has to be rebooted.

**Pull request merged**